viz/SonarVisualization.cpp

```diff
@@ -0,0 +1,176 @@
+#include "SonarVisualization.hpp"
+#include <osg/Geode>
+#include <base/Eigen.hpp>
+
+namespace vizkit3d
+{
+
+SonarVisualization::SonarVisualization():
+    motor_step(0),
+    num_steps(0),
+    new_sonar_scan(false),
+    full_scan(false)
+{
+    last_orientation.invalidate();
+    data.resize(1);
+}
+
+osg::ref_ptr< osg::Node > SonarVisualization::createMainNode()
+{
+    osg::ref_ptr<osg::Group> mainNode = new osg::Group();
+    pointGeom = new osg::Geometry;
+    pointsOSG = new osg::Vec3Array;
+    pointGeom->setVertexArray(pointsOSG);
+    color = new osg::Vec4Array;
+    pointGeom->setColorArray(color);
+    pointGeom->setColorBinding(osg::Geometry::BIND_PER_VERTEX);
+    pointGeom->getOrCreateStateSet()->setMode(GL_LIGHTING, osg::StateAttribute::OFF); 
+    drawArrays = new osg::DrawArrays( osg::PrimitiveSet::TRIANGLES, 0, pointsOSG->size() );
+    pointGeom->addPrimitiveSet(drawArrays.get());
+    
+    osg::ref_ptr<osg::Geode> geode = new osg::Geode;
+    geode->addDrawable(pointGeom.get());
+    mainNode->addChild(geode);
+    return mainNode;
+}
+
+/**
+ * Updates the class with the new sonar data of type SonarScan.
+ * 
+ * @param data new sonar data
+ */
+void SonarVisualization::updateDataIntern(const base::samples::Sonar& data)
+{
+    if(motor_step == 0)
+        return;
+    double current_pos = 0;
+    if(full_scan)
+        current_pos = ((data.bearings[0].rad+M_PI)/(2.0*M_PI))*num_steps;
+
+    SonarHelper bm;
+    bm.sonar_sample = data;
+    bm.orientation = last_orientation;
+    this->data[current_pos] = bm;
+    new_sonar_scan = true;
+}
+
+void SonarVisualization::updateDataIntern(const base::samples::RigidBodyState& data)
+{
+    last_orientation = data;
+}
+
+void SonarVisualization::setMotorStep(double motor_step)
+{
+    { 
+        boost::mutex::scoped_lock lockit(this->updateMutex);
+        if (motor_step == 0)
+            return; //Exception?
+        this->motor_step = motor_step;
+        this->num_steps =  (2.0*M_PI)/motor_step;
+        if(full_scan)
+            this->data.resize(num_steps+1);
+    }
+    emit propertyChanged("Motor Step");
+    setDirty();
+} 
+
+double SonarVisualization::getMotorStep()
+{
+    return motor_step;
+}
+
+void SonarVisualization::setFullScan(bool full_scan)
+{
+    { 
+        //boost::mutex::scoped_lock lockit(this->updateMutex);
+        this->full_scan = full_scan; 
+        data.clear();
+        if(!full_scan)
+        {
+            data.resize(1);    
+        }
+        else
+        {
+            if(motor_step != 0)
+            {
+                this->data.resize(num_steps+1);
+            }
+        }
+        
+    }
+    emit propertyChanged("Full Scan");
+    setDirty();
+}
+
+bool SonarVisualization::isFullScan()
+{
+    return full_scan;
+}
+
+void SonarVisualization::updateMainNode(osg::Node* node)
+{
+    if (new_sonar_scan)
```

Comment:
Please us an early return in this case, instead of having the whole method in a `if () ...` block

~~~
if (!new_sonar_scan)
   return;
~~~